Centralise code to map between UTF-8 and UTF-16 on Windows.

al45tair opened the pull request:
In various places we need to call the Windows API, and because Swift uses UTF-8 for its string representation, we can’t call the ANSI API functions (because the code page used for the ANSI functions varies depending on the system locale setting). Instead, we need to use the wide character APIs.

This means that we need to convert UTF-8 to wide character and vice-versa in various places in the runtime.

rdar://103397975

al45tair committed:
Centralise code to map between UTF-8 and UTF-16 on Windows.

In various places we need to call the Windows API, and because Swift uses UTF-8
for its string representation, we can’t call the ANSI API functions (because the
code page used for the ANSI functions varies depending on the system locale
setting). Instead, we need to use the wide character APIs.

This means that we need to convert UTF-8 to wide character and vice-versa in
various places in the runtime.

rdar://103397975

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
Since @grynspan and I both had essentially the same code in #62462 and #62577 respectively, it made sense to pull it out into a separate PR (before anyone else starts adding another copy to _their_ PR :-D).

While @compnerd suggested using the LLVM `ConvertUTF16ToUTF8` function, calling LLVM code from the runtime isn't possible unless we _copy_ the code into the LLVMSupport folder, and I'm keen that we minimise the amount of that that happens. Another alternative that wasn't mentioned is using the C++ library to do the conversion — but that causes allocation using `std::new` which is undesirable in the runtime (for the reason that `new` can be overridden by the user, which can create problems where user code gets called in contexts where some things might not work).

al45tair commented:
(Worth noting that presently these functions generate a fatal error when they fail. I'm not sure if that's desirable everywhere; maybe they shouldn't.)

mikeash reviewed (approved):
A fatal error makes sense when malloc fails. It doesn't seem right when the conversion fails, though.

compnerd commented on include/swift/Runtime/Win32.h:
```diff
@@ -0,0 +1,55 @@
+#include <wchar.h>
+
+namespace swift {
+namespace win32 {
```
I'm really not sure about the `win32` namespace here, this is pretty generic, and there are a few different implementations possible, some more portable than others.

al45tair replied:
This is only required for Windows, and only because of the wide character interfaces. For those, I think we want to use the Windows API to do the conversion, because that will be consistent with other things on the platform, rather than using some generic UTF-8 to UTF-16/UTF-16 to UTF-8 conversion code.

compnerd commented on include/swift/Runtime/Win32.h:
```diff
@@ -0,0 +1,55 @@
+///
+/// If @a str cannot be converted to UTF-16, a fatal error occurs.
+SWIFT_RUNTIME_STDLIB_INTERNAL
+wchar_t *copyWideFromUTF8(const char *str);
```
Can we call this `copyUTF16fromUTF8`?  It really is UTF16 encoded (even if you do something like wide is 32-bit).

al45tair replied:
I slightly prefer "Wide" in this context because naming it "UTF16" means people might think this was some kind of generic UTF-8 to UTF-16 interface, which it very much isn't. It's specifically here to convert Swift strings for the Windows wide character APIs.

I know that the Windows wide character encoding is in practice UTF-16 (mostly… sometimes it's UCS-2), but the goal here isn't to provide a general purpose character encoder.

compnerd commented on include/swift/Runtime/Win32.h:
```diff
@@ -0,0 +1,55 @@
+///
+/// If @a str cannot be converted to UTF-8, a fatal error occurs.
+SWIFT_RUNTIME_STDLIB_INTERNAL
+char *copyUTF8FromWide(const wchar_t *str);
```
Can we call this `copyUTF8FromUTF16`?

al45tair replied:
See previous comment :-)

compnerd reviewed (commented):
Does this really centralise it?  It seems that we should simultaneously update the symbolication codepath.

al45tair commented:
> Does this really centralise it? It seems that we should simultaneously update the symbolication codepath.

@grynspan is doing that in a separate PR. The point of this is to provide the functions in one place, in one PR.

al45tair committed:
Change to returning `nullptr` rather than triggering a fatal error.

Instead of triggering a fatal error on failure, return `nullptr` and
let the caller decide what to do about it.  The CommandLine code should
trigger a fatal error, of course.

rdar://103397975

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
> A fatal error makes sense when malloc fails. It doesn't seem right when the conversion fails, though.

Fixed it to return `nullptr` when it fails. `malloc()` failing in a recoverable way is in some sense more likely on Windows (it doesn't overcommit — at least, it isn't supposed to), so I think returning `nullptr` there too makes some sense.

grynspan commented:
> > A fatal error makes sense when malloc fails. It doesn't seem right when the conversion fails, though.
> 
> Fixed it to return `nullptr` when it fails. `malloc()` failing in a recoverable way is in some sense more likely on Windows (it doesn't overcommit — at least, it isn't supposed to), so I think returning `nullptr` there too makes some sense.

I think we should return `nullptr` and let the caller—oh, you just did that. Never mind!

grynspan commented on include/swift/Runtime/Win32.h:
```diff
@@ -0,0 +1,55 @@
+///
+/// If @a str cannot be converted to UTF-8, @c nullptr is returned.
+SWIFT_RUNTIME_STDLIB_INTERNAL
+char *copyUTF8FromWide(const wchar_t *str);
```
Elsewhere where we have Win32-specific symbols, I used `[_]swift_win32_` as a prefix. Although these symbols aren't going to be exported, you might want to adopt that prefix for consistency instead of using a `win32` nested namespace.

al45tair replied:
Since these are C++ symbols, I'm not sure why we would want to prefer C-style prefixes over C++-style namespaces. If we're exporting a C API, sure, putting `[_]swift_win32_` on the front would be the right choice.

al45tair replied:
In fact, we can't put them in a C++ namespace because `SWIFT_RUNTIME_STDLIB_INTERNAL` does `extern "C"`. So yes, we'll rename them, as you suggest.

grynspan commented on include/swift/Runtime/Win32.h:
```diff
@@ -0,0 +1,55 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_RUNTIME_WIN32_H
```
It might be worth moving `_swift_win32_withDbgHelpLibrary()` into this file as well. (Note that it _is_ necessary to continue to export that symbol because DbgHelp is a mess.)

al45tair replied:
I've moved that. It does mean that `Win32.h` will now bring in `windows.h`, as we need `HANDLE` for that.

grynspan commented on stdlib/public/CommandLineSupport/CommandLine.cpp:
```diff
@@ -244,7 +203,16 @@ static void swift::enumerateUnsafeArgv(const F& body) {
   if (LPWSTR *wargv = CommandLineToArgvW(GetCommandLineW(), &argc)) {
     std::for_each(wargv, wargv + argc, [=] (wchar_t *warg) {
-      auto arg = copyUTF8FromWide(warg);
+      auto arg = swift::win32::copyUTF8FromWide(warg);
```
Suggestion:

```c++
if (auto arg = swift::win32::copyUTF8FromWide(warg)) {
  body(argc, arg);
  free(arg);
} else {
  // Note that GetLastError() and errno may not be so useful here,
  // as in the error case we may have called free(), which might reset
  // either or both of them.
  swift::fatalError(0,
                      "Fatal error: Unable to convert argument '%ls' to "
                      "UTF-8: %lx, %d.\n",
                      warg, ::GetLastError(), errno);
}
```

Watch out for those 2-character tabs.

al45tair replied:
:-) If you use Emacs, which I do, the Swift sources configure Emacs automatically to do appropriate indentation.

al45tair committed:
Rename the functions as they're exported as C APIs.

`SWIFT_RUNTIME_STDLIB_INTERNAL` does `extern "C"`, so we can't put these
in a namespace and have to use a C-style prefix instead.

Also slightly rearrange the code in `CommandLine.cpp`.

rdar://103397975

al45tair committed:
Move _swift_win32_withDbgHelpLibrary into the new header.

It makes sense to move this function into the new Win32 header.

rdar://103397975

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test Windows platform

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
Make the UTF8-to-wide functions into SPI rather than internal.

They're useful functions to have; I don't think we want them to be API, but
having them as SPI is conceivably useful for other purposes, and avoids everyone
rolling their own copy.

rdar://103397975

al45tair commented:
@swift-ci Please smoke test

grynspan commented:
How would you feel about exposing `String.init(_wcs:)` and `String._asWCS(_:) rethrows` functions for slightly higher-level access to these interfaces? Obviously a permanent API would go through swift-evolution, but as an experimental/underbarred feature specific to Win32 interop, the utility of such interfaces cannot be understated.

al45tair commented:
> How would you feel about exposing String.init(_wcs:) and String._asWCS(_:) rethrows functions for slightly higher-level access to these interfaces?

That's more a question for the standard library folks, I think. I don't think that's necessarily the right thing anyway. IMO it'd be better to allow the Windows wide character APIs to be imported in a way such that we can pass `String`s to them directly, as we can with the narrow string APIs. But that's probably a discussion for elsewhere.

al45tair commented:
@swift-ci Please smoke test

grynspan commented:
> > How would you feel about exposing String.init(_wcs:) and String._asWCS(_:) rethrows functions for slightly higher-level access to these interfaces?
> 
> That's more a question for the standard library folks, I think. I don't think that's necessarily the right thing anyway. IMO it'd be better to allow the Windows wide character APIs to be imported in a way such that we can pass `String`s to them directly, as we can with the narrow string APIs. But that's probably a discussion for elsewhere.

Since `String(decodingCString: $0, as: UTF16.self)` does the thing already, I retracted my previous statement. :)
